Brilliant!

Looks like the recent changes is failing (differently):

```
======================================================================
ERROR: test_simple (corehq.apps.hqcase.tests.test_explode_cases.ExplodeCasesDbTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqcase/tests/test_explode_cases.py", line 131, in test_simple
    explode_cases(self.user_id, self.domain.name, 10)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqcase/tasks.py", line 16, in explode_cases
    keys = [[domain, owner_id, False] for owner_id in user.get_owner_ids()]
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1839, in get_owner_ids
    if self.project.locations_enabled and self.location:
AttributeError: 'dict' object has no attribute 'locations_enabled'
```

and one more similar one. Looks like domains are getting loaded as `dict`s instead of `Domain`s.

--- a/corehq/apps/domain/models.py
+++ b/corehq/apps/domain/models.py
@@ -18,6 +18,7 @@ from django.core.urlresolvers import reverse
 from django.db import models
 from django.utils.translation import ugettext_lazy as _
 from corehq.apps.appstore.models import SnapshotMixin
+from corehq.util.quickcache import QuickCache, quickcache
 from dimagi.utils.couch.cache import cache_core
 from dimagi.utils.couch.database import (
     iter_docs, get_db, get_safe_write_kwargs, apply_update, iter_bulk_delete
@@ -46,6 +47,19 @@ for lang in all_langs:
     lang_lookup[lang['three']] = lang['names'][0] # arbitrarily using the first name if there are multiple
     if lang['two'] != '':
         lang_lookup[lang['two']] = lang['names'][0]
+
+
+class _StrictQuickCache(QuickCache):
+    """
+    Just like QuickCache, but intercepts the function call to abort caching
+    under certain conditions
+    """
+    def __call__(self, *args, **kwargs):
+        strict = kwargs.get('strict', False)
+        if not strict:
+            return super(_StrictQuickCache, self).__call__(*args, **kwargs)
+        else:
+            return self.fn(*args, **kwargs)
 
 
 class DomainMigrations(DocumentSchema):
@@ -517,6 +531,7 @@ class Domain(Document, SnapshotMixin):
         return self.name
 
     @classmethod
+    @quickcache(['name'], timeout=30*60, helper_class=_StrictQuickCache)
     def get_by_name(cls, name, strict=False):
         if not name:
             # get_by_name should never be called with name as None (or '', etc)
@@ -534,16 +549,7 @@ class Domain(Document, SnapshotMixin):
                     notify_exception(None, '%r is not a valid domain name' % name)
                     return None
 
-        cache_key = _domain_cache_key(name)
-        if not strict:
-            MISSING = object()
-            res = cache.get(cache_key, MISSING)
-            if res != MISSING:
-                return res
-
         domain = cls._get_by_name(name, strict)
-        # 30 mins, so any unforeseen invalidation bugs aren't too bad.
-        cache.set(cache_key, domain, 30*60)
         return domain
 
     @classmethod
@@ -633,7 +639,7 @@ class Domain(Document, SnapshotMixin):
 
     def save(self, **params):
         super(Domain, self).save(**params)
-        cache.delete(_domain_cache_key(self.name))
+        self.get_by_name.clear(Domain, self.name)  # clear the domain cache
 
         from corehq.apps.domain.signals import commcare_domain_post_save
         results = commcare_domain_post_save.send_robust(sender='domain', domain=self)
@@ -1248,7 +1254,3 @@ class TransferDomainRequest(models.Model):
             'deactivate_url': self.deactivate_url(),
             'activate_url': self.activate_url(),
         }
-
-
-def _domain_cache_key(name):
-    return hashlib.md5(u'cchq:domain:{name}'.format(name=name).encode('utf-8')).hexdigest()

--- a/corehq/apps/domain/models.py
+++ b/corehq/apps/domain/models.py
@@ -552,28 +552,15 @@ class Domain(Document, SnapshotMixin):
                     notify_exception(None, '%r is not a valid domain name' % name)
                     return None
 
-        domain = cls._get_by_name(name, strict)
-        return domain
+        def _get_by_name(stale=False):
+            extra_args = {'stale': settings.COUCH_STALE_QUERY} if stale else {}
+            return cls.view("domain/domains", key=name, reduce=False, include_docs=True, **extra_args).first()
 
-    @classmethod
-    def _get_by_name(cls, name, strict=False):
-        extra_args = {'stale': settings.COUCH_STALE_QUERY} if not strict else {}
-
-        db = cls.get_db()
-        res = cache_core.cached_view(db, "domain/domains", key=name, reduce=False,
-                                     include_docs=True, wrapper=cls.wrap, force_invalidate=strict,
-                                     **extra_args)
-
-        if len(res) > 0:
-            result = res[0]
-        else:
-            result = None
-
-        if result is None and not strict:
+        domain = _get_by_name(stale=(not strict))
+        if domain is None and not strict:
             # on the off chance this is a brand new domain, try with strict
-            return cls.get_by_name(name, strict=True)
-
-        return result
+            domain = _get_by_name(stale=False)
+        return domain
 
     @classmethod
     def get_by_organization(cls, organization):

--- a/corehq/apps/domain/models.py
+++ b/corehq/apps/domain/models.py
@@ -59,7 +59,10 @@ class _StrictQuickCache(QuickCache):
         if not strict:
             return super(_StrictQuickCache, self).__call__(*args, **kwargs)
         else:
-            return self.fn(*args, **kwargs)
+            key = self.get_cache_key(*args, **kwargs)
+            content = self.fn(*args, **kwargs)
+            self.cache.set(key, content)
+            return content
 
 
 class DomainMigrations(DocumentSchema):

--- a/corehq/apps/reports/views.py
+++ b/corehq/apps/reports/views.py
@@ -655,7 +655,7 @@ def recalculate_hour(hour, hour_difference, minute_difference):
 
 
 def get_timezone_difference(domain):
-    return datetime.now(pytz.timezone(Domain._get_by_name(domain)['default_timezone'])).strftime('%z')
+    return datetime.now(pytz.timezone(Domain.get_by_name(domain)['default_timezone'])).strftime('%z')
 
 
 def calculate_day(interval, day, day_change):
@@ -733,7 +733,7 @@ def edit_scheduled_report(request, domain, scheduled_report_id=None,
     form.fields['recipient_emails'].choices = web_user_emails
 
     form.fields['hour'].help_text = "This scheduled report's timezone is %s (%s GMT)"  % \
-                                    (Domain._get_by_name(domain)['default_timezone'],
+                                    (Domain.get_by_name(domain)['default_timezone'],
                                     get_timezone_difference(domain)[:3] + ':' + get_timezone_difference(domain)[3:])
 
 

